### ISSUE TITLE
Generated Field Class Conflicts with Pydantic Field Class

### ISSUE THREAD

**Issue by gwukelic** [koxudaxi/datamodel-code-generator#1508]
**Describe the bug**

I have a schema that creates an invalid Python file. The issue seems to be that the `Field` class being generated conflicts with `Field` class from pydantic. How can I fix this?

```
error: Expression of type "Field" cannot be assigned to declared type "str"
    "Field" is incompatible with "str" (reportGeneralTypeIssues)
```

**To Reproduce**

Unacceptable schema ❌
```yaml
# input.yaml
components:
  schemas:
    BadSchema:
      type: object
      properties:
        fields:
          type: array
          items:
            oneOf:
            - type: object
              properties:
                a:
                  type: string
            - type: object
              properties:
                b:
                  type: string
                  pattern: "^[a-zA-Z_]+$"
```

What I run in the command line:
```sh
datamodel-codegen \
    --input input.yaml \
    --input-file-type "openapi" \
    --output output.py \
    --output-model-type pydantic.BaseModel \
    --field-constraints
```

Output

```python
# output.py using the unacceptable input.yaml file
from __future__ import annotations

from typing import List, Optional, Union

from pydantic import BaseModel, Field


class FieldModel(BaseModel):
    a: Optional[str] = None


class Field(BaseModel):
    b: Optional[str] = Field(None, regex='^[a-zA-Z_]+$')


class BadSchema(BaseModel):
    fields: Optional[List[Union[FieldModel, Field]]] = None
```

**Version:**
 - OS: macOS Ventura Version 13.3.1
 - datamodel-code-generator version: 0.21.4

**Comment by mjperrone** [comment]
@koxudaxi here's a test reproducing this issue https://github.com/koxudaxi/datamodel-code-generator/pull/1516

To run just the added test:

```bash
poetry run pytest tests/parser/test_openapi.py -k array_called_fields_with_oneOf_items
```

That should help give you a quick feedback loop if you if you wanted to work on fixing this.

I also might be able to attempt a fix if I find the time and feel confident enough. Koudai if you could help with some pointers to code I think I'd feel confident enough to try to implement this. Thanks

**Comment by koxudaxi** [comment]
@gwukelic @mjperrone @anis-campos 
Thank you for posting the comments.
We should check the defined name in the namespace.
I'm sorry I don't have the time recently. I would check it this week.

**Comment by mjperrone** [comment]
Thanks for the reply. If you don't end up having time to make the fix, I would appreciate a pointer in the code to where you think this logic would best be added so I can more confidently make some progress on this

**Comment by koxudaxi** [comment]
@mjperrone
Thank you for creating your PR and your help.
I have added [the logic](https://github.com/koxudaxi/datamodel-code-generator/pull/1516/files#diff-34a68196dba5b4f2cef091ddb835f794178254d4c94bff4a84f84e546f2f6db7R1023-R1049) to change the conflict class name to another name to your PR.
The fixes will show the result.
What do you think about it?
```python
from __future__ import annotations

from typing import List, Optional, Union

from pydantic import BaseModel, Field


class FieldModel(BaseModel):
    a: Optional[str] = None


class Field1(BaseModel):
    b: Optional[str] = Field(None, regex='^[a-zA-Z_]+$')


class BadSchema(BaseModel):
    fields: Optional[List[Union[FieldModel, Field1]]] = None
```

**Comment by mjperrone** [comment]
> What do you think about it?


I looked briefly at the code changes, and it makes some sense, but without digging deep to understand `ModelResolver` I don't think I can give much feedback on the actual code.

However, both in the test example provided and in my bigger spec which initially triggered this error, I am no longer seeing `class Field`, but `class Field1` and `class FieldModel`. 

I suppose the one suggestion I might make (which may already be implemented?) is to add a check to all tests making sure the generated code passes static type checking. Not sure how hard that is, or if it actually adds value since the test expectations are hopefully manually inspected and determined to not be invalid python...

Regardless, thank you so much for making this change! This is very helpful and I'm excited to keep using this cool library when this change gets released.

**Comment by mjperrone** [comment]
Is there any way I can help with the test failures that are happening on that PR branch? I'm excited to use this in a release

**Comment by mjperrone** [comment]
> I released the fixed version as `0.21.5` 

Thank you! I would consider this issue closed from my perspective.
